File: typescript/src/euclid.ts
import {IExtEuclid} from "./iexteuclid";

export class Euclid {

    static gcd(a: number, b: number): number {
        let tmp: number;
        while (b > 0) {
            tmp = b;
            b = a % b;
            a = tmp;
        }

        return a;
    }

    static extgcd(a: number, b: number): IExtEuclid {
        // Return x and y in form a*x + b*y
        // x is the inverse of a
        // y is the inverse if b
        // There is only an inverse if gcd(a,b) == 1
        if (a == 0) {
            const result: IExtEuclid = {x: 0, y: 1};
            return result;
        }else { 

            let result: IExtEuclid = this.extgcd(b % a, a);
            const ab = Math.floor(b / a);

            result.x = (result.x - ab * result.y);
            result.x = this.mod(result.x, b);

            return result;
        }
    }

    private static mod(a: number, b: number) {
        return ((a%b)+b) % b;
    }

    static moduloInverse(a: number, b: number): number {
        const result: IExtEuclid = this.extgcd(a, b);
        return result.x;
    }

}

File: C#/Euclid.cs
using System;

namespace Euclid {

    public class Euclid {

        private Euclid() {}

        public static int gcd(int a, int b) {
            
            int tmp;
            while(b > 0) {
                tmp = b;
                b = a % b;
                a = tmp;
            }

            return a;
        }

        public static (int x, int y) extgcd(int a, int b) {
            // Return x and y in form a*x + b*y
            // x is the inverse of a
            // y is the inverse if b
            // There is only an inverse if gcd(a,b) == 1
            if (a == 0) {
                return (0, 1);
            }else {
                var result = extgcd(b % a, a);
                int ab = (int)Math.Floor((float)b / (float)a);

                int x = result.y - ab * result.x;
                int y = result.x;

                return (x: x, y: y);
            }
        }

        private static int mod(int a, int b) => ((a%b)+b) % b; 

        public static int moduloInverse(int a, int b) {
            (int x, int y) = extgcd(a, b);
            return mod(x, b);
        }

    }

}